developer/src/kmc-generate/src/basic-generator.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Keyman is copyright (C) SIL International. MIT License.
+ *
+ * Basic generator -- common file generation functionality
+ */
+
+import { KeymanTargets } from "@keymanapp/common-types";
+import KEYMAN_VERSION from "@keymanapp/keyman-version";
+import { AbstractGenerator, GeneratorArtifacts } from "./abstract-generator.js";
+
+/**
+ * @internal
+ * Common functionality for generating projects. Do not instantiate
+ * this class, rather instantiate a subclass
+ */
+export class BasicGenerator extends AbstractGenerator {
+
+  // We'll generate a keyboard with a default 'en' locale if none is passed in
+  protected readonly DEFAULT_LOCALE = 'en';
+
+  protected templatePath: string;
+  protected languageTags: string[];
+  protected resolvedTargets: KeymanTargets.KeymanTarget[];
+
+  protected preGenerate() {
+    const dt = new Date();
+
+    this.languageTags = this.options.languageTags.length
+      ? this.options.languageTags.map(tag => new Intl.Locale(tag).minimize().toString())
+      : [this.DEFAULT_LOCALE];
+    //TODO-GENERATE: validate targets
+    if(this.options.targets.includes(KeymanTargets.KeymanTarget.any) || this.options.targets.length == 0) {
+      this.resolvedTargets = KeymanTargets.AllKeymanTargets.filter(target => target != KeymanTargets.KeymanTarget.any);
+    } else {
+      this.resolvedTargets = [].concat(this.options.targets);
+    }
```

Comment:
Should we put put languageTag related calls in a common wrapper/utility? (including the constant DEFAULT_LOCALE)

Reply:
I think I will do this as a separate refactor relating to BCP 47. There's a bunch of stuff around which BCP 47 dataset we use that will impact across the board.